src/EventStore.Projections.Core/Services/Management/ManagedProjection.cs

```diff
@@ -1022,7 +1030,7 @@ private readonly
 			if (_lastReplyEnvelope != null)
 				_lastReplyEnvelope.ReplyWith(new ProjectionManagementMessage.Updated(_name));
 			_lastReplyEnvelope = null;
-			if (Deleted) {
+			if (Deleted && !_pendingWritePersistedState) {
 				DisposeCoreProjection();
 				_output.Publish(new ProjectionManagementMessage.Internal.Deleted(_name, Id));
```

Comment:
is it true that if `_pendingWritePersistedState = true` here and we don't publish the `ProjectionManagementMessage.Internal.Deleted`  message, then we will always publish it later?

it doesn't seem to be necessarily the case when entering `StopUnlessPreparedOrLoaded()` in `Handle(ProjectionManagementMessage.Command.Delete message)`:
- for example, if the projection was already stopped and was in the `Prepared` or `Loaded` state, it seems that it'll just persist the state but never publish the `ProjectionManagementMessage.Internal.Deleted`  message.
- but if the projection was running, then it seems that it'll stop it, then do the deletion again here: https://github.com/EventStore/EventStore/blob/d3b31d85f7b8ba33be444141f01a2fe8973a18e8/src/EventStore.Projections.Core/Services/Management/ManagedProjection.cs#L564 which will eventually publish the event.

Reply:
So, [this statement](https://github.com/EventStore/EventStore/blob/2de22db4f53b48eb5ee2e7b4382615eed172c1c9/src/EventStore.Projections.Core/Services/Management/ManagedProjection.cs#L422) ensures that projection state will always be Stopped or Faulted
[Delete](https://github.com/EventStore/EventStore/blob/2de22db4f53b48eb5ee2e7b4382615eed172c1c9/src/EventStore.Projections.Core/Services/Management/ManagedProjection.cs#L445) ensures that _pendingWritePersistedState will be true
From step 1 and 2, it means [WritePersistedState](https://github.com/EventStore/EventStore/blob/2de22db4f53b48eb5ee2e7b4382615eed172c1c9/src/EventStore.Projections.Core/Services/Management/ManagedProjection.cs#L721) will always be called
Hence, _pendingWritePersistedState will always be set to false when write completes

Reply:
@hunter1703 thanks for the clarifications, i get it now: `Reply()` is always being called twice: once from `DeleteIfConditionsAreMet()` and once from `StartOrLoadStopped()`.

The check: `Deleted && !_pendingWritePersistedState` ensures that the `ProjectionManagementMessage.Internal.Deleted` message will be published only once from whichever process (i.e the stream deletions or the persisted state write) finishes last.

Reply:
another thought (not directly linked to the problem this PR is solving)

there are currently many possible states on startup due to the parallelism of the stream deletions/write of the persisted state and it makes it difficult to ensure we're actually handling all these cases correctly (list may not be exhaustive):
<Deleting = true, Deleted = false, projection deleted from $projections-$all>
<Deleting = false, Deleted = true, projection deleted from $projections-$all>
<Deleting = true, Deleted = false, projection NOT deleted from $projections-$all> (partial deletion)
<Deleting = false, Deleted = true, projection NOT deleted from $projections-$all> (partial deletion)
<Deleting = false, Deleted = false, projection NOT deleted from $projections-$all> (e.g some projection streams deleted)

in the future, we could consider serializing the operations as follows so that it's simpler to reason about it / process partial deletions on startup:
1. write persisted state with `Deleting = true, Deleted = false`
2. delete the projection streams
3. write persisted state with `Deleting = false, Deleted = true`
4. delete the projection from $projections-$all

Reply:
Ah. okay. I can look into this as part of a separate PR. For now, I am resolving this thread. Thanks for the comments.